File: metadata-ingestion/src/datahub/specific/dataprocessinstance.py
# ABOUTME: Patch builder for DataProcessInstance entity.
# ABOUTME: Supports adding/removing inputs, outputs, and their edges.
from typing import List, Optional, Union

from datahub.emitter.mcp_patch_builder import MetadataPatchProposal
from datahub.metadata.schema_classes import (
    DataProcessInstanceInputClass,
    DataProcessInstanceOutputClass,
    EdgeClass as Edge,
    KafkaAuditHeaderClass,
    SystemMetadataClass,
)
from datahub.metadata.urns import Urn


class DataProcessInstancePatchBuilder(
    MetadataPatchProposal,
):
    """
    Patch builder for DataProcessInstance entities.

    Supports patching:
    - inputs (URN array in dataProcessInstanceInput aspect)
    - inputEdges (Edge array in dataProcessInstanceInput aspect)
    - outputs (URN array in dataProcessInstanceOutput aspect)
    - outputEdges (Edge array in dataProcessInstanceOutput aspect)

    Example usage:
        builder = DataProcessInstancePatchBuilder("urn:li:dataProcessInstance:my-instance")
        builder.add_input_edge("urn:li:dataset:(urn:li:dataPlatform:hive,input_table,PROD)")
        builder.add_output_edge("urn:li:dataset:(urn:li:dataPlatform:hive,output_table,PROD)")

        for mcp in builder.build():
            emitter.emit(mcp)
    """

    def __init__(
        self,
        urn: str,
        system_metadata: Optional[SystemMetadataClass] = None,
        audit_header: Optional[KafkaAuditHeaderClass] = None,
    ) -> None:
        """
        Initializes a DataProcessInstancePatchBuilder instance.

        Args:
            urn: The URN of the data process instance.
            system_metadata: The system metadata (optional).
            audit_header: The Kafka audit header (optional).
        """
        super().__init__(
            urn, system_metadata=system_metadata, audit_header=audit_header
        )

    # ==================== Input Operations ====================

    def add_input(
        self, input_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Adds an input URN to the inputs array.

        Args:
            input_urn: The input URN (dataset, mlModel, etc.)

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(input_urn)
        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "add",
            path=("inputs", urn_str),
            value=urn_str,
        )
        return self

    def remove_input(
        self, input_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Removes an input URN from the inputs array.

        Args:
            input_urn: The input URN to remove.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(input_urn)
        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "remove",
            path=("inputs", urn_str),
            value={},
        )
        return self

    def set_inputs(self, inputs: List[str]) -> "DataProcessInstancePatchBuilder":
        """
        Sets the inputs array, replacing any existing inputs.

        Args:
            inputs: A list of input URN strings.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "add",
            path=("inputs",),
            value=inputs,
        )
        return self

    # ==================== Input Edge Operations ====================

    def add_input_edge(
        self, input_edge: Union[Edge, Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Adds an input edge to the inputEdges array.

        Args:
            input_edge: The input edge, which can be an Edge object, Urn, or string.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        if isinstance(input_edge, Edge):
            input_urn: str = input_edge.destinationUrn
            edge: Edge = input_edge
        elif isinstance(input_edge, (Urn, str)):
            input_urn = str(input_edge)
            edge = Edge(destinationUrn=input_urn)

        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "add",
            path=("inputEdges", input_urn),
            value=edge,
        )
        return self

    def remove_input_edge(
        self, input_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Removes an input edge from the inputEdges array.

        Args:
            input_urn: The destination URN of the edge to remove.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(input_urn)
        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "remove",
            path=("inputEdges", urn_str),
            value={},
        )
        return self

    def set_input_edges(self, input_edges: List[Edge]) -> "DataProcessInstancePatchBuilder":
        """
        Sets the inputEdges array, replacing any existing input edges.

        Args:
            inputs: A list of Edge objects representing the input edges.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        self._add_patch(
            DataProcessInstanceInputClass.ASPECT_NAME,
            "add",
            path=("inputEdges",),
            value=input_edges,
        )
        return self

    # ==================== Output Operations ====================

    def add_output(
        self, output_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Adds an output URN to the outputs array.

        Args:
            output_urn: The output URN (dataset, mlModel, etc.)

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(output_urn)
        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "add",
            path=("outputs", urn_str),
            value=urn_str,
        )
        return self

    def remove_output(
        self, output_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Removes an output URN from the outputs array.

        Args:
            output_urn: The output URN to remove.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(output_urn)
        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "remove",
            path=("outputs", urn_str),
            value={},
        )
        return self

    def set_outputs(self, outputs: List[str]) -> "DataProcessInstancePatchBuilder":
        """
        Sets the outputs array, replacing any existing outputs.

        Args:
            outputs: A list of output URN strings.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "add",
            path=("outputs",),
            value=outputs,
        )
        return self

    # ==================== Output Edge Operations ====================

    def add_output_edge(
        self, output_edge: Union[Edge, Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Adds an output edge to the outputEdges array.

        Args:
            output_edge: The output edge, which can be an Edge object, Urn, or string.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        if isinstance(output_edge, Edge):
            output_urn: str = output_edge.destinationUrn
            edge: Edge = output_edge
        elif isinstance(output_edge, (Urn, str)):
            output_urn = str(output_edge)
            edge = Edge(destinationUrn=output_urn)

        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "add",
            path=("outputEdges", output_urn),
            value=edge,
        )
        return self

    def remove_output_edge(
        self, output_urn: Union[Urn, str]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Removes an output edge from the outputEdges array.

        Args:
            output_urn: The destination URN of the edge to remove.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        urn_str = str(output_urn)
        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "remove",
            path=("outputEdges", urn_str),
            value={},
        )
        return self

    def set_output_edges(
        self, outputs: List[Edge]
    ) -> "DataProcessInstancePatchBuilder":
        """
        Sets the outputEdges array, replacing any existing output edges.

        Args:
            outputs: A list of Edge objects representing the output edges.

        Returns:
            The DataProcessInstancePatchBuilder instance.
        """
        self._add_patch(
            DataProcessInstanceOutputClass.ASPECT_NAME,
            "add",
            path=("outputEdges",),
            value=outputs,
        )
        return self

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/v2/entity/DataProcessInstance.java
// ABOUTME: Represents a DataHub DataProcessInstance entity with fluent builder API.
// ABOUTME: Supports patch-based updates for input/output lineage operations.
package datahub.client.v2.entity;

import com.linkedin.common.Edge;
import com.linkedin.common.urn.Urn;
import com.linkedin.data.template.RecordTemplate;
import com.linkedin.dataprocess.DataProcessInstanceInput;
import com.linkedin.dataprocess.DataProcessInstanceOutput;
import com.linkedin.dataprocess.DataProcessInstanceProperties;
import com.linkedin.metadata.aspect.patch.builder.DataProcessInstanceInputPatchBuilder;
import com.linkedin.metadata.aspect.patch.builder.DataProcessInstanceOutputPatchBuilder;
import datahub.client.v2.annotations.RequiresMutable;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import lombok.extern.slf4j.Slf4j;

/**
 * Represents a DataHub DataProcessInstance entity with fluent builder API.
 *
 * <p>A DataProcessInstance represents a specific execution/run of a data processing job (e.g., an
 * Airflow task run, a Spark job execution). Unlike DataJob which represents the definition of a
 * job, DataProcessInstance captures the actual execution with its runtime lineage.
 *
 * <p>This implementation uses patch-based updates that accumulate until save(). All mutations
 * (addInput, addOutput) create patch MCPs rather than modifying aspects directly.
 *
 * <p>Example usage:
 *
 * <pre>
 * DataProcessInstance instance = DataProcessInstance.builder()
 *     .id("my-job-run-12345")
 *     .name("ETL Pipeline Run #12345")
 *     .build();
 *
 * // Add input/output lineage (creates patches)
 * instance.addInputEdge("urn:li:dataset:(urn:li:dataPlatform:hive,raw_data,PROD)");
 * instance.addOutputEdge("urn:li:dataset:(urn:li:dataPlatform:hive,processed_data,PROD)");
 *
 * // Save to DataHub (emits accumulated patches)
 * client.entities().upsert(instance);
 * </pre>
 *
 * @see Entity
 */
@Slf4j
public class DataProcessInstance extends Entity {

  private static final String ENTITY_TYPE = "dataProcessInstance";

  /**
   * Constructs a new DataProcessInstance entity.
   *
   * @param urn the data process instance URN
   */
  private DataProcessInstance(@Nonnull Urn urn) {
    super(urn);
  }

  /**
   * Constructs a DataProcessInstance loaded from server.
   *
   * @param urn the data process instance URN
   * @param aspects current aspects from server
   */
  private DataProcessInstance(@Nonnull Urn urn, @Nonnull Map<String, RecordTemplate> aspects) {
    super(urn, aspects);
  }

  /**
   * Copy constructor for creating mutable copy via .mutable().
   *
   * @param other the DataProcessInstance to copy from
   */
  protected DataProcessInstance(@Nonnull DataProcessInstance other) {
    super(
        other.urn,
        other.cache,
        other.client,
        other.mode,
        new HashMap<>(),
        new ArrayList<>(),
        new HashMap<>(),
        false,
        false);
  }

  @Override
  @Nonnull
  public String getEntityType() {
    return ENTITY_TYPE;
  }

  @Override
  @Nonnull
  public List<Class<? extends RecordTemplate>> getDefaultAspects() {
    return List.of(
        com.linkedin.common.Status.class,
        DataProcessInstanceProperties.class,
        DataProcessInstanceInput.class,
        DataProcessInstanceOutput.class);
  }

  @Override
  @Nonnull
  public DataProcessInstance mutable() {
    if (!readOnly) {
      return this;
    }
    return new DataProcessInstance(this);
  }

  /**
   * Creates a new builder for DataProcessInstance.
   *
   * @return a new builder instance
   */
  @Nonnull
  public static Builder builder() {
    return new Builder();
  }

  // ==================== Name Operations ====================

  /**
   * Gets the name for this data process instance (lazy-loaded).
   *
   * @return the name, or null if not set
   */
  @Nullable
  public String getName() {
    DataProcessInstanceProperties props = getAspectLazy(DataProcessInstanceProperties.class);
    return props != null ? props.getName() : null;
  }

  // ==================== Input Lineage Operations ====================

  /**
   * Adds an input URN to this data process instance.
   *
   * <p>Input URNs represent assets that this instance reads from. This is the simple form that
   * stores just the URN.
   *
   * @param inputUrn the input URN (e.g., dataset, mlModel)
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addInput(@Nonnull String inputUrn) {
    checkNotReadOnly("add input");
    try {
      Urn urn = Urn.createFromString(inputUrn);
      DataProcessInstanceInputPatchBuilder patch =
          new DataProcessInstanceInputPatchBuilder().urn(getUrn()).addInput(urn);
      addPatchMcp(patch.build());
      log.debug("Added input patch: {}", inputUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + inputUrn, e);
    }
    return this;
  }

  /**
   * Adds an input URN to this data process instance.
   *
   * @param inputUrn the input URN
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addInput(@Nonnull Urn inputUrn) {
    checkNotReadOnly("add input");
    DataProcessInstanceInputPatchBuilder patch =
        new DataProcessInstanceInputPatchBuilder().urn(getUrn()).addInput(inputUrn);
    addPatchMcp(patch.build());
    log.debug("Added input patch: {}", inputUrn);
    return this;
  }

  /**
   * Removes an input URN from this data process instance.
   *
   * @param inputUrn the input URN to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeInput(@Nonnull String inputUrn) {
    checkNotReadOnly("remove input");
    try {
      Urn urn = Urn.createFromString(inputUrn);
      DataProcessInstanceInputPatchBuilder patch =
          new DataProcessInstanceInputPatchBuilder().urn(getUrn()).removeInput(urn);
      addPatchMcp(patch.build());
      log.debug("Added remove input patch: {}", inputUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + inputUrn, e);
    }
    return this;
  }

  /**
   * Removes an input URN from this data process instance.
   *
   * @param inputUrn the input URN to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeInput(@Nonnull Urn inputUrn) {
    checkNotReadOnly("remove input");
    DataProcessInstanceInputPatchBuilder patch =
        new DataProcessInstanceInputPatchBuilder().urn(getUrn()).removeInput(inputUrn);
    addPatchMcp(patch.build());
    log.debug("Added remove input patch: {}", inputUrn);
    return this;
  }

  /**
   * Gets the inputs for this data process instance (lazy-loaded).
   *
   * @return list of input URNs, or empty list if not set
   */
  @Nonnull
  public List<Urn> getInputs() {
    DataProcessInstanceInput input = getAspectLazy(DataProcessInstanceInput.class);
    if (input == null || input.getInputs() == null) {
      return Collections.emptyList();
    }
    return new ArrayList<>(input.getInputs());
  }

  // ==================== Input Edge Lineage Operations ====================

  /**
   * Adds an input edge to this data process instance.
   *
   * <p>Input edges represent assets that this instance reads from, with additional metadata. This
   * creates lineage from the input asset to this instance.
   *
   * @param destinationUrn the destination URN of the edge
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addInputEdge(@Nonnull String destinationUrn) {
    checkNotReadOnly("add input edge");
    try {
      Urn urn = Urn.createFromString(destinationUrn);
      DataProcessInstanceInputPatchBuilder patch =
          new DataProcessInstanceInputPatchBuilder().urn(getUrn()).addInputEdge(urn);
      addPatchMcp(patch.build());
      log.debug("Added input edge patch: {}", destinationUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + destinationUrn, e);
    }
    return this;
  }

  /**
   * Adds an input edge to this data process instance.
   *
   * @param destinationUrn the destination URN of the edge
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addInputEdge(@Nonnull Urn destinationUrn) {
    checkNotReadOnly("add input edge");
    DataProcessInstanceInputPatchBuilder patch =
        new DataProcessInstanceInputPatchBuilder().urn(getUrn()).addInputEdge(destinationUrn);
    addPatchMcp(patch.build());
    log.debug("Added input edge patch: {}", destinationUrn);
    return this;
  }

  /**
   * Adds an input edge with full Edge properties to this data process instance.
   *
   * @param edge the edge to add
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addInputEdge(@Nonnull Edge edge) {
    checkNotReadOnly("add input edge");
    DataProcessInstanceInputPatchBuilder patch =
        new DataProcessInstanceInputPatchBuilder().urn(getUrn()).addInputEdge(edge);
    addPatchMcp(patch.build());
    log.debug("Added input edge patch: {}", edge.getDestinationUrn());
    return this;
  }

  /**
   * Removes an input edge from this data process instance.
   *
   * @param destinationUrn the destination URN of the edge to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeInputEdge(@Nonnull String destinationUrn) {
    checkNotReadOnly("remove input edge");
    try {
      Urn urn = Urn.createFromString(destinationUrn);
      DataProcessInstanceInputPatchBuilder patch =
          new DataProcessInstanceInputPatchBuilder().urn(getUrn()).removeInputEdge(urn);
      addPatchMcp(patch.build());
      log.debug("Added remove input edge patch: {}", destinationUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + destinationUrn, e);
    }
    return this;
  }

  /**
   * Removes an input edge from this data process instance.
   *
   * @param destinationUrn the destination URN of the edge to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeInputEdge(@Nonnull Urn destinationUrn) {
    checkNotReadOnly("remove input edge");
    DataProcessInstanceInputPatchBuilder patch =
        new DataProcessInstanceInputPatchBuilder().urn(getUrn()).removeInputEdge(destinationUrn);
    addPatchMcp(patch.build());
    log.debug("Added remove input edge patch: {}", destinationUrn);
    return this;
  }

  /**
   * Gets the input edges for this data process instance (lazy-loaded).
   *
   * @return list of input edges, or empty list if not set
   */
  @Nonnull
  public List<Edge> getInputEdges() {
    DataProcessInstanceInput input = getAspectLazy(DataProcessInstanceInput.class);
    if (input == null || input.getInputEdges() == null) {
      return Collections.emptyList();
    }
    return new ArrayList<>(input.getInputEdges());
  }

  /**
   * Gets the input edge destination URNs for this data process instance (lazy-loaded).
   *
   * @return list of destination URNs, or empty list if not set
   */
  @Nonnull
  public List<Urn> getInputEdgeUrns() {
    return getInputEdges().stream().map(Edge::getDestinationUrn).collect(Collectors.toList());
  }

  // ==================== Output Lineage Operations ====================

  /**
   * Adds an output URN to this data process instance.
   *
   * <p>Output URNs represent assets that this instance writes to. This is the simple form that
   * stores just the URN.
   *
   * @param outputUrn the output URN (e.g., dataset, mlModel)
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addOutput(@Nonnull String outputUrn) {
    checkNotReadOnly("add output");
    try {
      Urn urn = Urn.createFromString(outputUrn);
      DataProcessInstanceOutputPatchBuilder patch =
          new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).addOutput(urn);
      addPatchMcp(patch.build());
      log.debug("Added output patch: {}", outputUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + outputUrn, e);
    }
    return this;
  }

  /**
   * Adds an output URN to this data process instance.
   *
   * @param outputUrn the output URN
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addOutput(@Nonnull Urn outputUrn) {
    checkNotReadOnly("add output");
    DataProcessInstanceOutputPatchBuilder patch =
        new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).addOutput(outputUrn);
    addPatchMcp(patch.build());
    log.debug("Added output patch: {}", outputUrn);
    return this;
  }

  /**
   * Removes an output URN from this data process instance.
   *
   * @param outputUrn the output URN to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeOutput(@Nonnull String outputUrn) {
    checkNotReadOnly("remove output");
    try {
      Urn urn = Urn.createFromString(outputUrn);
      DataProcessInstanceOutputPatchBuilder patch =
          new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).removeOutput(urn);
      addPatchMcp(patch.build());
      log.debug("Added remove output patch: {}", outputUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + outputUrn, e);
    }
    return this;
  }

  /**
   * Removes an output URN from this data process instance.
   *
   * @param outputUrn the output URN to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeOutput(@Nonnull Urn outputUrn) {
    checkNotReadOnly("remove output");
    DataProcessInstanceOutputPatchBuilder patch =
        new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).removeOutput(outputUrn);
    addPatchMcp(patch.build());
    log.debug("Added remove output patch: {}", outputUrn);
    return this;
  }

  /**
   * Gets the outputs for this data process instance (lazy-loaded).
   *
   * @return list of output URNs, or empty list if not set
   */
  @Nonnull
  public List<Urn> getOutputs() {
    DataProcessInstanceOutput output = getAspectLazy(DataProcessInstanceOutput.class);
    if (output == null || output.getOutputs() == null) {
      return Collections.emptyList();
    }
    return new ArrayList<>(output.getOutputs());
  }

  // ==================== Output Edge Lineage Operations ====================

  /**
   * Adds an output edge to this data process instance.
   *
   * <p>Output edges represent assets that this instance writes to, with additional metadata. This
   * creates lineage from this instance to the output asset.
   *
   * @param destinationUrn the destination URN of the edge
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addOutputEdge(@Nonnull String destinationUrn) {
    checkNotReadOnly("add output edge");
    try {
      Urn urn = Urn.createFromString(destinationUrn);
      DataProcessInstanceOutputPatchBuilder patch =
          new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).addOutputEdge(urn);
      addPatchMcp(patch.build());
      log.debug("Added output edge patch: {}", destinationUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + destinationUrn, e);
    }
    return this;
  }

  /**
   * Adds an output edge to this data process instance.
   *
   * @param destinationUrn the destination URN of the edge
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addOutputEdge(@Nonnull Urn destinationUrn) {
    checkNotReadOnly("add output edge");
    DataProcessInstanceOutputPatchBuilder patch =
        new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).addOutputEdge(destinationUrn);
    addPatchMcp(patch.build());
    log.debug("Added output edge patch: {}", destinationUrn);
    return this;
  }

  /**
   * Adds an output edge with full Edge properties to this data process instance.
   *
   * @param edge the edge to add
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance addOutputEdge(@Nonnull Edge edge) {
    checkNotReadOnly("add output edge");
    DataProcessInstanceOutputPatchBuilder patch =
        new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).addOutputEdge(edge);
    addPatchMcp(patch.build());
    log.debug("Added output edge patch: {}", edge.getDestinationUrn());
    return this;
  }

  /**
   * Removes an output edge from this data process instance.
   *
   * @param destinationUrn the destination URN of the edge to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeOutputEdge(@Nonnull String destinationUrn) {
    checkNotReadOnly("remove output edge");
    try {
      Urn urn = Urn.createFromString(destinationUrn);
      DataProcessInstanceOutputPatchBuilder patch =
          new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).removeOutputEdge(urn);
      addPatchMcp(patch.build());
      log.debug("Added remove output edge patch: {}", destinationUrn);
    } catch (URISyntaxException e) {
      throw new IllegalArgumentException("Invalid URN: " + destinationUrn, e);
    }
    return this;
  }

  /**
   * Removes an output edge from this data process instance.
   *
   * @param destinationUrn the destination URN of the edge to remove
   * @return this data process instance
   */
  @RequiresMutable
  @Nonnull
  public DataProcessInstance removeOutputEdge(@Nonnull Urn destinationUrn) {
    checkNotReadOnly("remove output edge");
    DataProcessInstanceOutputPatchBuilder patch =
        new DataProcessInstanceOutputPatchBuilder().urn(getUrn()).removeOutputEdge(destinationUrn);
    addPatchMcp(patch.build());
    log.debug("Added remove output edge patch: {}", destinationUrn);
    return this;
  }

  /**
   * Gets the output edges for this data process instance (lazy-loaded).
   *
   * @return list of output edges, or empty list if not set
   */
  @Nonnull
  public List<Edge> getOutputEdges() {
    DataProcessInstanceOutput output = getAspectLazy(DataProcessInstanceOutput.class);
    if (output == null || output.getOutputEdges() == null) {
      return Collections.emptyList();
    }
    return new ArrayList<>(output.getOutputEdges());
  }

  /**
   * Gets the output edge destination URNs for this data process instance (lazy-loaded).
   *
   * @return list of destination URNs, or empty list if not set
   */
  @Nonnull
  public List<Urn> getOutputEdgeUrns() {
    return getOutputEdges().stream().map(Edge::getDestinationUrn).collect(Collectors.toList());
  }

  // ==================== Builder ====================

  /** Builder for creating DataProcessInstance entities with a fluent API. */
  public static class Builder {
    private String id;
    private String name;
    private Urn urn;

    /**
     * Sets the instance ID (required if URN not provided).
     *
     * <p>This ID uniquely identifies the instance. It could be a run ID, execution ID, or any
     * unique identifier.
     *
     * @param id the instance ID
     * @return this builder
     */
    @Nonnull
    public Builder id(@Nonnull String id) {
      this.id = id;
      return this;
    }

    /**
     * Sets the display name for this instance.
     *
     * @param name the display name
     * @return this builder
     */
    @Nonnull
    public Builder name(@Nullable String name) {
      this.name = name;
      return this;
    }

    /**
     * Sets the URN for this instance directly, skipping ID-based URN generation.
     *
     * <p>Use this when you have an existing URN (e.g., from server fetch).
     *
     * @param urn the data process instance URN
     * @return this builder
     */
    @Nonnull
    public Builder urn(@Nonnull Urn urn) {
      this.urn = urn;
      return this;
    }

    /**
     * Sets the URN for this instance from a string.
     *
     * @param urn the data process instance URN string
     * @return this builder
     * @throws IllegalArgumentException if the URN string is invalid
     */
    @Nonnull
    public Builder urn(@Nonnull String urn) {
      try {
        this.urn = Urn.createFromString(urn);
      } catch (URISyntaxException e) {
        throw new IllegalArgumentException("Invalid URN: " + urn, e);
      }
      return this;
    }

    /**
     * Builds the DataProcessInstance entity.
     *
     * @return the DataProcessInstance entity
     * @throws IllegalArgumentException if required fields are missing
     */
    @Nonnull
    public DataProcessInstance build() {
      try {
        Urn finalUrn;

        if (urn != null) {
          finalUrn = urn;
        } else if (id != null) {
          finalUrn = Urn.createFromString("urn:li:dataProcessInstance:" + id);
        } else {
          throw new IllegalArgumentException("Either id or urn is required");
        }

        DataProcessInstance instance = new DataProcessInstance(finalUrn);

        // Cache DataProcessInstanceProperties aspect if name is provided
        if (name != null) {
          DataProcessInstanceProperties props = new DataProcessInstanceProperties();
          props.setName(name);
          // created is a required field
          props.setCreated(
              new com.linkedin.common.AuditStamp()
                  .setTime(System.currentTimeMillis())
                  .setActor(
                      Urn.createFromString("urn:li:corpuser:__datahub_system_client_user__")));

          String aspectName = instance.getAspectName(DataProcessInstanceProperties.class);
          instance.cache.put(aspectName, props, AspectSource.LOCAL, true);
        }

        return instance;
      } catch (URISyntaxException e) {
        throw new IllegalArgumentException("Failed to create data process instance URN", e);
      }
    }
  }
}
